File: agent-typescript/src/mcp.ts
/**
 * MCP wiring: one Client per (Slack user, MCP server).
 *
 * connectUser() never throws for ordinary failures. Servers that answer
 * 401 move to "needs-auth" and carry the Keycard authorization URL the
 * MCP SDK produced (discovery per RFC 9728, Dynamic Client Registration
 * per RFC 7591, then the PKCE authorization request). The bot posts that
 * link; when the user's browser comes back to /oauth/callback,
 * completeAuthorization() exchanges the code and reconnects.
 */

import { randomBytes } from "node:crypto";
import { UnauthorizedError } from "@modelcontextprotocol/sdk/client/auth.js";
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StreamableHTTPClientTransport } from "@modelcontextprotocol/sdk/client/streamableHttp.js";
import type { ServerEntry } from "./config.js";
import { SlackUserOAuthProvider } from "./oauth-provider.js";
import type { FileAuthStore } from "./storage.js";

export type SessionStatus = "ready" | "needs-auth" | "failed";

export interface ServerSession {
	serverKey: string;
	serverUrl: string;
	status: SessionStatus;
	client: Client | undefined;
	/** Present when status is "needs-auth". */
	authorizationUrl: URL | undefined;
}

export type ReadySession = ServerSession & { client: Client };

export function isReady(session: ServerSession): session is ReadySession {
	return session.status === "ready" && session.client !== undefined;
}

interface PendingAuthorization {
	userId: string;
	serverKey: string;
}

export class McpManager {
	private readonly sessionsByUser = new Map<string, Map<string, ServerSession>>();
	/** OAuth state parameter -> who started the flow. One-time use. */
	private readonly pendingStates = new Map<string, PendingAuthorization>();

	constructor(
		private readonly servers: ServerEntry[],
		private readonly store: FileAuthStore,
		private readonly redirectUri: string,
	) {}

	/**
	 * Get (or create) this user's sessions, connecting any that are not
	 * ready yet. Repeat calls are cheap: ready sessions are left alone.
	 */
	async connectUser(userId: string): Promise<ServerSession[]> {
		let sessions = this.sessionsByUser.get(userId);
		if (!sessions) {
			sessions = new Map();
			this.sessionsByUser.set(userId, sessions);
		}
		for (const server of this.servers) {
			const existing = sessions.get(server.key);
			if (existing && isReady(existing)) {
				continue;
			}
			sessions.set(server.key, await this.connectServer(userId, server));
		}
		return [...sessions.values()];
	}

	/**
	 * Handle the browser hitting /oauth/callback. Resolves the state back
	 * to (user, server), exchanges the authorization code for tokens via
	 * the MCP SDK (transport.finishAuth), and reconnects the session.
	 */
	async completeAuthorization(
		state: string,
		code: string,
	): Promise<PendingAuthorization> {
		const pending = this.pendingStates.get(state);
		if (!pending) {
			throw new Error("Unknown or expired authorization state.");
		}
		this.pendingStates.delete(state);

		const server = this.servers.find((entry) => entry.key === pending.serverKey);
		if (!server) {
			throw new Error(`No configured MCP server named ${pending.serverKey}.`);
		}

		// finishAuth only performs the token exchange; the provider already
		// holds the client registration and PKCE verifier in the file store,
		// so a fresh transport works. Tokens are persisted by the provider.
		const provider = this.buildProvider(pending.userId, server, () => {});
		const transport = new StreamableHTTPClientTransport(new URL(server.url), {
			authProvider: provider,
		});
		await transport.finishAuth(code);

		const session = await this.connectServer(pending.userId, server);
		let sessions = this.sessionsByUser.get(pending.userId);
		if (!sessions) {
			sessions = new Map();
			this.sessionsByUser.set(pending.userId, sessions);
		}
		sessions.set(server.key, session);

		if (!isReady(session)) {
			throw new Error(
				`Authorized ${server.key}, but reconnecting failed (status: ${session.status}).`,
			);
		}
		return pending;
	}

	/**
	 * Drop a stale token and re-run the OAuth flow for one server.
	 *
	 * Called when a tool fails because the user's grant was revoked or expired
	 * upstream. The catch is that the session to the MCP server is still
	 * healthy (its own token is valid), so it never re-challenges on its own —
	 * only the delegated token exchange the MCP server does at call time fails.
	 *
	 * We delete just this server's stored token (keeping the DCR registration)
	 * so the next connect gets a fresh 401 -> authorization flow, which
	 * re-consents the revoked grant. Returns the reconnected session; when it
	 * comes back "needs-auth" it carries the fresh authorization URL to post.
	 */
	async forceReauth(userId: string, serverKey: string): Promise<ServerSession> {
		const server = this.servers.find((entry) => entry.key === serverKey);
		if (!server) {
			throw new Error(`No configured MCP server named ${serverKey}.`);
		}

		if (this.store.clearTokens(userId, serverKey)) {
			console.log(`Cleared stored token for ${userId}/${serverKey}; forcing re-auth`);
		} else {
			// No token to clear means the next connect won't get a 401 and we'd
			// post a "reconnect" link that cleared nothing. Surface it instead
			// of silently claiming recovery.
			console.warn(`forceReauth: no stored token found for ${userId}/${serverKey}`);
		}

		const session = await this.connectServer(userId, server);
		let sessions = this.sessionsByUser.get(userId);
		if (!sessions) {
			sessions = new Map();
			this.sessionsByUser.set(userId, sessions);
		}
		sessions.set(serverKey, session);
		return session;
	}

	private buildProvider(
		userId: string,
		server: ServerEntry,
		onAuthorizationUrl: (url: URL) => void,
	): SlackUserOAuthProvider {
		return new SlackUserOAuthProvider(userId, server.key, this.store, this.redirectUri, {
			issueState: () => {
				const state = randomBytes(24).toString("base64url");
				this.pendingStates.set(state, { userId, serverKey: server.key });
				return state;
			},
			onAuthorizationUrl,
		});
	}

	private async connectServer(
		userId: string,
		server: ServerEntry,
	): Promise<ServerSession> {
		const authUrlBox: { url: URL | undefined } = { url: undefined };
		const provider = this.buildProvider(userId, server, (url) => {
			authUrlBox.url = url;
		});
		const client = new Client({ name: "tutorial-slack-agent", version: "0.1.0" });
		const transport = new StreamableHTTPClientTransport(new URL(server.url), {
			authProvider: provider,
		});

		const session: ServerSession = {
			serverKey: server.key,
			serverUrl: server.url,
			status: "failed",
			client: undefined,
			authorizationUrl: undefined,
		};

		try {
			// If the server answers 401, the MCP SDK runs the full pre-auth
			// flow inside connect(): discovery, DCR, PKCE, and a call to our
			// provider's redirectToAuthorization, then throws UnauthorizedError.
			await client.connect(transport);
			session.status = "ready";
			session.client = client;
		} catch (error) {
			await transport.close().catch(() => {});
			if (error instanceof UnauthorizedError) {
				session.status = "needs-auth";
				session.authorizationUrl = authUrlBox.url;
			} else {
				console.error(
					`Failed to connect to MCP server ${server.key} (${server.url}):`,
					error instanceof Error ? error.message : error,
				);
			}
		}
		return session;
	}
}

File: agent-python/src/slack_agent/mcp.py
"""keycardai-mcp wiring: one ClientManager, one client per Slack user.

The keycardai-mcp SDK does the heavy lifting for Keycard-protected MCP
servers. On a 401 it discovers the protected-resource metadata (RFC 9728),
registers itself as a public OAuth client via Dynamic Client Registration
(RFC 7591), and runs the authorization-code flow with PKCE. No client
secret is stored in this app; the MCP server side holds the confidential
credentials for its upstream providers.

Three pieces:

- SQLiteBackend persists per-user tokens and pending auth state on disk,
  so users only authorize once per server.
- StarletteAuthCoordinator handles the OAuth redirect. It does not run its
  own HTTP server; we mount its completion endpoint on our Starlette app
  at /oauth/callback (see main.py).
- ClientManager hands out one Client per context id. We use
  "slack:<user_id>" so every Slack user gets isolated token storage.
"""

from __future__ import annotations

import logging

from keycardai.mcp.client import Client, ClientManager, SQLiteBackend, StarletteAuthCoordinator
from keycardai.mcp.client.types import AuthChallenge

from slack_agent.config import ServerEntry

logger = logging.getLogger(__name__)


def build_manager(
    servers: list[ServerEntry],
    redirect_uri: str,
    db_path: str,
) -> tuple[ClientManager, StarletteAuthCoordinator]:
    """Create the shared ClientManager and its auth coordinator."""
    backend = SQLiteBackend(db_path)
    coordinator = StarletteAuthCoordinator(backend=backend, redirect_uri=redirect_uri)
    manager = ClientManager(
        servers={
            entry.key: {"url": entry.url, "auth": {"type": "oauth"}}
            for entry in servers
        },
        auth_coordinator=coordinator,
    )
    return manager, coordinator


def context_id_for(slack_user_id: str) -> str:
    return f"slack:{slack_user_id}"


def slack_user_from_context(context_id: str) -> str | None:
    """Inverse of context_id_for. Returns None for non-Slack contexts."""
    if context_id.startswith("slack:"):
        return context_id.removeprefix("slack:")
    return None


async def get_user_client(manager: ClientManager, slack_user_id: str) -> Client:
    """Get (or create) this user's MCP client and connect it.

    connect() never raises for ordinary failures. Sessions that need OAuth
    move to an auth-pending state; check client.get_auth_challenges()
    afterwards to see if the user must authorize. Repeat calls are cheap:
    already-healthy sessions are left alone.
    """
    client = await manager.get_client(context_id_for(slack_user_id))
    await client.connect()
    return client


async def force_reauth(client: Client, server_name: str) -> list[AuthChallenge]:
    """Drop a stale token and re-run the OAuth flow for one server.

    Called when a tool fails because the user's grant was revoked or expired
    upstream. The catch is that the session to the MCP server is still
    healthy (its own token is valid), so it never re-challenges on its own —
    only the delegated token exchange the MCP server does at call time fails.

    We delete just this server's stored token so the next request has no
    bearer, forcing a fresh 401 -> authorization flow (which re-consents the
    revoked grant), then reconnect and return the new challenge(s) to post.
    This is what lets the demo recover without deleting the token database.
    """
    oauth_storage = (
        client.context.storage_path()
        .for_server(server_name)
        .for_connection()
        .for_oauth()
        .build()
    )
    cleared = await oauth_storage.delete("tokens")
    if cleared:
        logger.info("Cleared stored token for %s; forcing re-auth", server_name)
    else:
        # No token at the expected key means the storage layout drifted: the
        # reconnect below won't get a 401 and we'd post a "reconnect" link that
        # cleared nothing. Surface it instead of silently claiming recovery.
        logger.warning(
            "force_reauth: no stored token found for %s at %s — storage layout "
            "may have changed",
            server_name,
            oauth_storage._namespace,
        )
    await client.connect(server_name, force_reconnect=True)
    return await client.get_auth_challenges(server_name)
